File: apps/backend-python/src/housing_platform/payments/finalization.py
import uuid
from datetime import UTC, datetime
from typing import Any

from sqlalchemy import select
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.orm import Session

from housing_platform.auth.errors import (
    BadRequestError,
    BookingExpiredError,
    NotFoundError,
    PaymentAmountMismatchError,
)
from housing_platform.db.models import Bookings, PaymentEvents, Payments


class PaymentFinalizationService:
    def __init__(self, db: Session) -> None:
        self._db = db

    def finalize_successful_payment(
        self,
        *,
        order_id: uuid.UUID,
        payment_key: str,
        amount_krw: int,
        toss_response: dict[str, Any] | None,
    ) -> Payments:
        payment = self._db.scalar(
            select(Payments).where(Payments.order_id == order_id).with_for_update()
        )
        if payment is None:
            raise NotFoundError("Payment not found")

        if payment.status == "confirmed":
            return payment

        if payment.amount_krw != amount_krw:
            raise PaymentAmountMismatchError("Payment amount mismatch")

        booking = self._db.scalar(
            select(Bookings).where(Bookings.id == payment.booking_id).with_for_update()
        )
        if booking is None:
            raise NotFoundError("Booking not found")

        if booking.status == "confirmed":
            payment.status = "confirmed"
            payment.payment_key = payment_key
            if toss_response is not None:
                payment.toss_response = toss_response
            if payment.confirmed_at is None:
                payment.confirmed_at = datetime.now(UTC)
            self._db.flush()
            return payment

        if booking.status != "pending_payment":
            raise BadRequestError("Booking is not awaiting payment")

        now = datetime.now(UTC)
        if booking.hold_expires_at is not None and booking.hold_expires_at <= now:
            booking.status = "expired"
            self._db.flush()
            raise BookingExpiredError("Booking hold has expired")

        payment.status = "confirmed"
        payment.payment_key = payment_key
        payment.toss_response = toss_response
        payment.confirmed_at = now
        booking.status = "confirmed"
        self._db.flush()
        return payment

    def mark_payment_failed(
        self,
        *,
        order_id: uuid.UUID,
        reason: str | None,
        toss_response: dict[str, Any] | None,
    ) -> Payments:
        payment = self._db.scalar(
            select(Payments).where(Payments.order_id == order_id).with_for_update()
        )
        if payment is None:
            raise NotFoundError("Payment not found")

        if payment.status == "confirmed":
            return payment

        trimmed_reason = reason.strip() if reason else None
        payment.status = "failed"
        payment.failed_reason = trimmed_reason if trimmed_reason else None
        if toss_response is not None:
            payment.toss_response = toss_response

        booking = self._db.scalar(select(Bookings).where(Bookings.id == payment.booking_id))
        if booking is not None and booking.status in ("pending_payment", "payment_failed"):
            booking.status = "payment_failed"

        self._db.flush()
        return payment

    def record_payment_event(
        self,
        *,
        event_id: str,
        payment_id: uuid.UUID,
        booking_id: uuid.UUID,
        event_type: str,
        payload: dict[str, Any],
    ) -> PaymentEvents | None:
        existing = self._db.scalar(
            select(PaymentEvents).where(PaymentEvents.event_id == event_id)
        )
        if existing is not None:
            return existing

        stmt = (
            insert(PaymentEvents)
            .values(
                event_id=event_id,
                payment_id=payment_id,
                booking_id=booking_id,
                event_type=event_type,
                payload=payload,
            )
            .on_conflict_do_nothing(index_elements=[PaymentEvents.event_id])
            .returning(PaymentEvents)
        )
        inserted = self._db.scalar(stmt)
        if inserted is not None:
            return inserted

        return self._db.scalar(select(PaymentEvents).where(PaymentEvents.event_id == event_id))

File: apps/backend-go/internal/payments/finalization.go
package payments

import (
	"context"
	"encoding/json"
	"errors"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"

	apierrors "github.com/housing-platform/backend-go/internal/api/errors"
	"github.com/housing-platform/backend-go/internal/db"
)

type FinalizationService struct{}

func NewFinalizationService() *FinalizationService {
	return &FinalizationService{}
}

func (s *FinalizationService) FinalizeSuccessfulPayment(
	ctx context.Context,
	q pgx.Tx,
	orderID pgtype.UUID,
	paymentKey string,
	amountKrw int,
	tossResponse map[string]any,
) (*db.Payment, error) {
	var payment db.Payment
	err := q.QueryRow(ctx, `
		SELECT id, order_id, payment_key, booking_id, customer_id, amount_krw, status,
		       toss_response, failed_reason, confirmed_at, created_at, updated_at
		FROM public.payments WHERE order_id = $1 FOR UPDATE
	`, orderID).Scan(
		&payment.ID, &payment.OrderID, &payment.PaymentKey, &payment.BookingID,
		&payment.CustomerID, &payment.AmountKrw, &payment.Status, &payment.TossResponse,
		&payment.FailedReason, &payment.ConfirmedAt, &payment.CreatedAt, &payment.UpdatedAt,
	)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			panic(apierrors.NotFound("Payment not found"))
		}
		return nil, err
	}

	if payment.Status == db.PaymentStatusConfirmed {
		return &payment, nil
	}

	if int(payment.AmountKrw) != amountKrw {
		panic(apierrors.PaymentAmountMismatch("Payment amount mismatch"))
	}

	var booking db.Booking
	err = q.QueryRow(ctx, `
		SELECT id, status, hold_expires_at FROM public.bookings WHERE id = $1 FOR UPDATE
	`, payment.BookingID).Scan(&booking.ID, &booking.Status, &booking.HoldExpiresAt)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			panic(apierrors.NotFound("Booking not found"))
		}
		return nil, err
	}

	tossJSON := marshalJSON(tossResponse)
	now := time.Now().UTC()

	if booking.Status == db.BookingStatusConfirmed {
		_, err = q.Exec(ctx, `
			UPDATE public.payments
			SET status = 'confirmed', payment_key = $2, toss_response = $3,
			    confirmed_at = COALESCE(confirmed_at, $4)
			WHERE order_id = $1
		`, orderID, paymentKey, tossJSON, now)
		if err != nil {
			return nil, err
		}
		payment.Status = db.PaymentStatusConfirmed
		payment.PaymentKey = &paymentKey
		return &payment, nil
	}

	if booking.Status != db.BookingStatusPendingPayment {
		panic(apierrors.BadRequest("Booking is not awaiting payment"))
	}

	if booking.HoldExpiresAt.Valid && !booking.HoldExpiresAt.Time.After(now) {
		_, _ = q.Exec(ctx, `UPDATE public.bookings SET status = 'expired' WHERE id = $1`, payment.BookingID)
		panic(apierrors.BookingExpired("Booking hold has expired"))
	}

	_, err = q.Exec(ctx, `
		UPDATE public.payments
		SET status = 'confirmed', payment_key = $2, toss_response = $3, confirmed_at = $4
		WHERE order_id = $1
	`, orderID, paymentKey, tossJSON, now)
	if err != nil {
		return nil, err
	}
	_, err = q.Exec(ctx, `UPDATE public.bookings SET status = 'confirmed' WHERE id = $1`, payment.BookingID)
	if err != nil {
		return nil, err
	}

	payment.Status = db.PaymentStatusConfirmed
	payment.PaymentKey = &paymentKey
	return &payment, nil
}

func (s *FinalizationService) MarkPaymentFailed(
	ctx context.Context,
	q pgx.Tx,
	orderID pgtype.UUID,
	reason string,
	tossResponse map[string]any,
) (*db.Payment, error) {
	var payment db.Payment
	err := q.QueryRow(ctx, `
		SELECT id, order_id, booking_id, status
		FROM public.payments WHERE order_id = $1 FOR UPDATE
	`, orderID).Scan(&payment.ID, &payment.OrderID, &payment.BookingID, &payment.Status)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			panic(apierrors.NotFound("Payment not found"))
		}
		return nil, err
	}

	if payment.Status == db.PaymentStatusConfirmed {
		return &payment, nil
	}

	trimmed := strings.TrimSpace(reason)
	var failedReason *string
	if trimmed != "" {
		failedReason = &trimmed
	}

	tossJSON := marshalJSON(tossResponse)
	_, err = q.Exec(ctx, `
		UPDATE public.payments
		SET status = 'failed', failed_reason = $2, toss_response = $3
		WHERE order_id = $1
	`, orderID, failedReason, tossJSON)
	if err != nil {
		return nil, err
	}

	_, _ = q.Exec(ctx, `
		UPDATE public.bookings SET status = 'payment_failed'
		WHERE id = $1 AND status IN ('pending_payment', 'payment_failed')
	`, payment.BookingID)

	payment.Status = db.PaymentStatusFailed
	return &payment, nil
}

func (s *FinalizationService) RecordPaymentEvent(
	ctx context.Context,
	q pgx.Tx,
	eventID string,
	paymentID, bookingID pgtype.UUID,
	eventType string,
	payload map[string]any,
) error {
	payloadJSON := marshalJSON(payload)
	_, err := q.Exec(ctx, `
		INSERT INTO public.payment_events (event_id, payment_id, booking_id, event_type, payload)
		VALUES ($1, $2, $3, $4, $5)
		ON CONFLICT (event_id) DO NOTHING
	`, eventID, paymentID, bookingID, eventType, payloadJSON)
	return err
}

func marshalJSON(value map[string]any) []byte {
	if value == nil {
		return nil
	}
	b, _ := json.Marshal(value)
	return b
}
